src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,370 @@
+                if (request != null)
+                {
+                    HttpWebResponse response = s_httpResponseAccessor(request);
+                    if (response != null)
+                    {
+                        s_instance.RaiseResponseEvent(request, response);
+                    }
+                }
+
+                base.RemoveAt(index);
+            }
+        }
+
+#endregion
+
+#region private methods
+
+        /// <summary>
+        /// Private constructor. Make sure every caller has to create this object via the Initialize
+        /// method.
+        /// </summary>
+        private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
+        {
+        }
+
+        private void RaiseRequestEvent(HttpWebRequest request)
+        {
+            Guid loggingRequestId = Guid.Empty;
+
+            // If System.Net.Http.Request is on, raise the event
+            if (this.IsEnabled(RequestWriteName))
+            {
+                long timestamp = Stopwatch.GetTimestamp();
+                loggingRequestId = Guid.NewGuid();
+                this.Write(RequestWriteName,
+                    new
+                    {
+                        Request = request,
+                        //LoggingRequestId = loggingRequestId,
+                        Timestamp = timestamp
+                    }
+                );
+            }
+        }
+
+        private void RaiseResponseEvent(HttpWebRequest request, HttpWebResponse response)
+        {
+            Guid loggingRequestId = Guid.Empty;
+
+            // TODO: Figure what's the right way to detect error conditions here.
+            if (response.StatusCode >= (HttpStatusCode)400)
+            {
+                if (this.IsEnabled(ExceptionEventName))
+                {
+                    //this.Write(ExceptionEventName, new { Exception = ex, Request = request });
+                    this.Write(ExceptionEventName, new { Response = response });
+                }
+            }
+
+            if (this.IsEnabled(ResponseWriteName))
+            {
+                long timestamp = Stopwatch.GetTimestamp();
+                this.Write(ResponseWriteName,
+                    new
+                    {
+                        Request = request,
+                        Response = response,
+                        //LoggingRequestId = loggingRequestId,
+                        TimeStamp = timestamp
+                        //RequestTaskStatus = responseTask.Status
+                    }
+                );
+            }
+        }
+
+        private static void PrepareReflectionObjects()
+        {
+            // At any point, if the operation failed, it should just throw, including NullReferenceException.
+            // The caller should catch all exceptions and swallow.
+
+            // First step: Get all the reflection objects we will ever need.
+            Assembly systemNetHttpAssembly = typeof(ServicePoint).Assembly;
```

Comment:
/cc @weshaggard, @davidsh, @karelz 

There are several concerns here (aside from the unsigned CLA):
1. The solution is based on reflection over .Net Framework internals that are not guaranteed to exist in the next versions.
2. The code is separated from the .Net Framework code which is not public and so a test showing that System.Net changes may affect it is not possible.
3. There are no tests to verify this even works.